{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mummer pipeline \n",
    "\n",
    "**Versions**  \n",
    "Mummer: 4.0.0beta2  \n",
    "Python: 3.7.3  \n",
    "GNUPLOT: 4.6.0  \n",
    "ghostscript: 9.22  \n",
    "Mercator: 2013.01.11"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mercator\n",
    "\n",
    "Run the following perl script to identify blocks of synteny:  \n",
    "Run_mercator.pl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/perl -w\n",
    "\n",
    "use strict;\n",
    "\n",
    "my $ref = shift;\n",
    "my $query = shift;\n",
    "\n",
    "$ref =~ m/(.+)\\.fa/;\n",
    "my $refBase = $1;\n",
    "$query =~ m/(.+)\\.fa/;\n",
    "my $queryBase = $1;\n",
    "\n",
    "#Repeatmasker\n",
    "# Mask low complexity repeats\n",
    "my $refLow = $ref.\".low\";\n",
    "my $refLowMasked = $refLow.\".masked\";\n",
    "\n",
    "my $queryLow = $query.\".low\";\n",
    "my $queryLowMasked = $queryLow.\".masked\";\n",
    "\n",
    "system(\"ln -s $ref $refLow\");\n",
    "system(\"ln -s $query $queryLow\");\n",
    "system(\"RepeatMasker -e ncbi -pa 28 -q -no_is -noint -species drosophila $refLow >& RM.refLOW.log\");\n",
    "system(\"RepeatMasker -e ncbi -pa 28 -q -no_is -noint -species drosophila $queryLow >& RM.queryLOW.log\");\n",
    "\n",
    "# Mask interspersed repeats\n",
    "my $refInt = $ref.\".int\";\n",
    "my $refIntMasked = $refInt.\".masked\";\n",
    "\n",
    "my $queryInt = $query.\".int\";\n",
    "my $queryIntMasked = $queryInt.\".masked\";\n",
    "\n",
    "system(\"ln -s $ref $refInt\");\n",
    "system(\"ln -s $query $queryInt\");\n",
    "system(\"RepeatMasker -e ncbi -pa 28 -q -no_is -nolow -species drosophila $refInt >& RM.refINT.log\");\n",
    "system(\"RepeatMasker -e ncbi -pa 28 -q -no_is -nolow -species drosophila $queryInt >& RM.queryINT.log\");\n",
    "\n",
    "die unless -s $refLowMasked;\n",
    "die unless -s $queryLowMasked;\n",
    "die unless -s $refIntMasked;\n",
    "die unless -s $queryIntMasked;\n",
    "\n",
    "# Merge masking into one hardmasked file\n",
    "my $refMasked = $ref.\".masked\";\n",
    "my $queryMasked = $query.\".masked\";\n",
    "\n",
    "print \"nmerge $refIntMasked $refLowMasked > $refMasked\\n\";\n",
    "print \"nmerge $queryIntMasked $queryLowMasked > $queryMasked\\n\";\n",
    "system(\"nmerge $refIntMasked $refLowMasked > $refMasked\");\n",
    "system(\"nmerge $queryIntMasked $queryLowMasked > $queryMasked\");\n",
    "\n",
    "die unless -s $refMasked;\n",
    "die unless -s $queryMasked;\n",
    "\n",
    "# Create softmasked file\n",
    "my $refSoftMasked = $ref.\".softmasked\";\n",
    "my $querySoftMasked = $query.\".softmasked\";\n",
    "system(\"faSoftMask $ref $refMasked > $refSoftMasked\");\n",
    "system(\"faSoftMask $query $queryMasked > $querySoftMasked\");\n",
    "\n",
    "die unless -s $refSoftMasked;\n",
    "die unless -s $querySoftMasked;\n",
    "\n",
    "#Create SDB files\n",
    "my $refSDB = $refBase.\".sdb\";\n",
    "my $querySDB = $queryBase.\".sdb\";\n",
    "system(\"cat $refSoftMasked | fa2sdb -c $refSDB\");\n",
    "system(\"cat $querySoftMasked | fa2sdb -c $querySDB\");\n",
    "\n",
    "die unless -s $refSDB;\n",
    "die unless -s $querySDB;\n",
    "\n",
    "# Run SNAP with fly HMM\n",
    "my $refZFF = $refBase.\".zff\";\n",
    "my $refGFF = $refBase.\".gff\";\n",
    "my $queryZFF = $queryBase.\".zff\";\n",
    "my $queryGFF = $queryBase.\".gff\";\n",
    "system(\"runSnap /home/cee53/pkg/snap/HMM/fly < $refIntMasked > $refZFF\");\n",
    "system(\"runSnap /home/cee53/pkg/snap/HMM/fly < $queryIntMasked > $queryZFF\");\n",
    "\n",
    "die unless -s $refZFF;\n",
    "die unless -s $queryZFF;\n",
    "\n",
    "system(\"cat $refZFF | zff2gtf --source=SNAP > $refGFF\");\n",
    "system(\"cat $queryZFF | zff2gtf --source=SNAP > $queryGFF\");\n",
    "\n",
    "die unless -s $refGFF;\n",
    "die unless -s $queryGFF;\n",
    "\n",
    "my $tree = \"($refBase\\:0.05,$queryBase\\:0.05)\\;\";\n",
    "\n",
    "# Create input files for Mercator\n",
    "system(\"makeMercatorInput $refBase $queryBase\");\n",
    "\n",
    "# Make treefile\n",
    "system(\"echo \\'$tree\\' > treefile\");\n",
    "\n",
    "#Move files into input dir\n",
    "system(\"mkdir input\");\n",
    "system(\"mkdir output\");\n",
    "system(\"mv $refBase\\* input/.\");\n",
    "system(\"mv $queryBase\\* input/.\");\n",
    "system(\"mv treefile input/.\");\n",
    "\n",
    "#Constructing Orthology Map\n",
    "system(\"mercator -i input -o output $refBase $queryBase\");\n",
    "\n",
    "#Move to output dir\n",
    "chdir(\"output\");\n",
    "system(\"ln -s ../input/*.sdb .\");\n",
    "system(\"ln -s ../input/treefile\");\n",
    "\n",
    "#Breakpoint Finding\n",
    "# Convert the orthology map into a more general homology map\n",
    "system(\"omap2hmap genomes < pre.map > pre.h.map\");\n",
    "# Create the graph relating the breakpoint regions\n",
    "system(\"makeBreakpointGraph pre.h.map treefile\");\n",
    "# Make pairwise alignments for breakpoint regions\n",
    "system(\"mkdir bp_alignments\");\n",
    "system(\"makeBreakpointAlignmentInput --out-dir=bp_alignments\");\n",
    "system(\"mavidAlignDirs --init-dir=bp_alignments\");\n",
    "# Find a good configuration of breakpoints\n",
    "system(\"findBreakpoints --resolution 1000 pre.h.map treefile edges bp_alignments > breakpoints\");\n",
    "# Refine the map by splitting the breakpoint regions\n",
    "system(\"breakMap breakpoints < pre.h.map > better.h.map\");\n",
    "# Convert back to the orthology map format\n",
    "system(\"hmap2omap genomes < better.h.map > better.map\");\n",
    "\n",
    "#Generate MAVID input\n",
    "#Convert pairwise hits to alignment constraints\n",
    "system(\"phits2constraints -i ../input < pairwisehits > constraints\");\n",
    "#Create directories and files for alignment\n",
    "system(\"mkdir alignments\");\n",
    "system(\"makeAlignmentInput --map=better.map . alignments\");\n",
    "\n",
    "#Align with MAVID\n",
    "#Align all sequence files in directory structure\n",
    "system(\"mavidAlignDirs --init-dir=alignments\");"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Determine which _D. triauraria_ megascaffold corresponds to which Muller element based on chromosome count table from <tt>Genome_Assembly<tt> pipeline. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**_D. triauraria_**  \n",
    "\n",
    "|Muller element | Scaffold |\n",
    "|--- | ---\n",
    "Muller_ A|HiC_Scaffold_5|\n",
    "Muller_B |HiC_scaffold_1|\n",
    "Muller_C |HiC_scaffold_2|\n",
    "Muller_D |HiC_scaffold_7|\n",
    "Muller_E |HiC_scaffold_8|\n",
    "Muller_F |HiC_scaffold_6|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run the python script called <tt> extract_scaff_rename_muller.py <tt>  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Python script to extract scaffolds and rename as Muller element. Can reverse complement scaffolds as well if necessary.\n",
    "\n",
    "instructions for \"extract_scaff_rename_muller.py\"\n",
    "Script is to rename HiC scaffolds from Drosophila species to appropriate Muller elements and reverse complement the scaffold if necessary.\n",
    "\n",
    "Be sure to input two output files, the first will print the Muller elements in the order they are found and the second will be the reordered Muller elements (A,B,C,D,E,F).\n",
    "\n",
    "in bash shell type:\n",
    "python extract_scaff_rename_muller.py -h \n",
    "\n",
    "Output will be \n",
    "Options:\n",
    "  -h, --help            show this help message and exit\n",
    "  -i FASTA              scaffolds.fasta\n",
    "  -o OUTFILE            output1.fasta\n",
    "  -t OUTFILE2, --o2=OUTFILE2\n",
    "                        output2.fasta - this will be the correctly ordered\n",
    "                        file\n",
    "  -a MA                 scaffold corresp. to Muller A\n",
    "  -b MB                 scaffold corresp. to Muller B\n",
    "  -c MC                 scaffold corresp. to Muller C\n",
    "  -d MD                 scaffold corresp. to Muller D\n",
    "  -e ME                 scaffold corresp. to Muller E\n",
    "  -f MF                 scaffold corresp. to Muller F\n",
    "  -u MAR, --rcMA=MAR    scaffold corresp. to Muller A - use to reverse\n",
    "                        complement scaff\n",
    "  -v MBR, --rcMB=MBR    scaffold corresp. to Muller B - use to reverse\n",
    "                        complement scaff\n",
    "  -w MCR, --rcMC=MCR    scaffold corresp. to Muller C - use to reverse\n",
    "                        complement scaff\n",
    "  -x MDR, --rcMD=MDR    scaffold corresp. to Muller D - use to reverse\n",
    "                        complement scaff\n",
    "  -y MER, --rcME=MER    scaffold corresp. to Muller E - use to reverse\n",
    "                        complement scaff\n",
    "  -z MFR, --rcMF=MFR    scaffold corresp. to Muller F - use to reverse\n",
    "                        complement scaff\n",
    "\n",
    "\n",
    "After running the script, input OUTFILE2 name into the <OUTFILE2> slot in the nucmer_slurm.sh script in the folder. Update prefix for nucmer output files, the folder for the SLURM output, and the nucmer parameters as necessary.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#/projects/genetics/ellison_lab/nicole/mummer_projects/script/\n",
    "from optparse import OptionParser\n",
    "from Bio import SeqIO\n",
    "\n",
    "parser = OptionParser()\n",
    "parser.add_option(\"-i\", dest=\"fasta\", type=\"string\", action=\"store\", help = \"scaffolds.fasta\")\n",
    "parser.add_option(\"-o\", dest=\"outfile\", type = \"string\", action = \"store\", help = \"output1.fasta\") \n",
    "parser.add_option(\"-t\",\"--o2\", dest=\"outfile2\", type = \"string\", action = \"store\", help = \"output2.fasta - this will be the correctly ordered file\")\n",
    "parser.add_option(\"-a\", dest=\"MA\", type = \"string\", help = \"scaffold corresp. to Muller A\")\n",
    "parser.add_option(\"-b\", dest=\"MB\", type = \"string\", help = \"scaffold corresp. to Muller B\")\n",
    "parser.add_option(\"-c\", dest=\"MC\", type = \"string\", help = \"scaffold corresp. to Muller C\")\n",
    "parser.add_option(\"-d\", dest=\"MD\", type = \"string\", help = \"scaffold corresp. to Muller D\")\n",
    "parser.add_option(\"-e\", dest=\"ME\", type = \"string\", help = \"scaffold corresp. to Muller E\")\n",
    "parser.add_option(\"-f\", dest=\"MF\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller F\")\n",
    "parser.add_option(\"-u\",\"--rcMA\", dest=\"MAR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller A - use to reverse complement scaff\")\n",
    "parser.add_option(\"-v\", \"--rcMB\", dest=\"MBR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller B - use to reverse complement scaff\")\n",
    "parser.add_option(\"-w\", \"--rcMC\", dest=\"MCR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller C - use to reverse complement scaff\")\n",
    "parser.add_option(\"-x\", \"--rcMD\", dest=\"MDR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller D - use to reverse complement scaff\")\n",
    "parser.add_option(\"-y\", \"--rcME\", dest=\"MER\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller E - use to reverse complement scaff\")\n",
    "parser.add_option(\"-z\", \"--rcMF\", dest=\"MFR\", type = \"string\", action = \"store\", help = \"scaffold corresp. to Muller F - use to reverse complement scaff\")\n",
    "\n",
    "(options, args) = parser.parse_args()\n",
    "f = open(options.outfile, 'w')\n",
    "\n",
    "for record in SeqIO.parse(options.fasta, \"fasta\"):\n",
    "\tif record.id == options.MA:\n",
    "\t\tMA = record\t\n",
    "\t\tf.write(\">\" + \"Muller_A\" + '\\n' + str(MA.seq) + '\\n')\n",
    "\telif record.id == options.MB:\n",
    "\t\tMB = record\n",
    "\t\tf.write(\">\" + \"Muller_B\" + '\\n' + str(MB.seq) + '\\n')\n",
    "\telif record.id == options.MC:\n",
    "\t\tMC = record\n",
    "\t\tf.write(\">\" + \"Muller_C\" + '\\n' + str(MC.seq) + '\\n')\n",
    "\telif record.id == options.MD:\n",
    "\t\tMD = record\n",
    "\t\tf.write(\">\" + \"Muller_D\" + '\\n' + str(MD.seq) + '\\n')\n",
    "\telif record.id == options.ME:\n",
    "\t\tME = record\n",
    "\t\tf.write(\">\" + \"Muller_E\" + '\\n' + str(ME.seq) + '\\n')\n",
    "\telif record.id == options.MF:\n",
    "\t\tMF = record\n",
    "\t\tf.write(\">\" + \"Muller_F\" + '\\n' + str(MF.seq) + '\\n')\n",
    "\n",
    "for record in SeqIO.parse(options.fasta, \"fasta\"):\n",
    "        if record.id == options.MAR:\n",
    "                MAR = record\n",
    "                f.write(\">\" + \"Muller_A\" + '\\n' + str(MAR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MBR:\n",
    "                MBR = record\n",
    "                f.write(\">\" + \"Muller_B\" + '\\n' + str(MBR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MCR:\n",
    "                MCR = record\n",
    "                f.write(\">\" + \"Muller_C\" + '\\n' + str(MCR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MDR:\n",
    "                MDR = record\n",
    "                f.write(\">\" + \"Muller_D\" + '\\n' + str(MDR.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MER:\n",
    "                MER = record\n",
    "                f.write(\">\" + \"Muller_E\" + '\\n' + str(MER.seq.reverse_complement()) + '\\n')\n",
    "        elif record.id == options.MFR:\n",
    "                MFR = record\n",
    "                f.write(\">\" + \"Muller_F\" + '\\n' + str(MFR.seq.reverse_complement()) + '\\n')\n",
    "\n",
    "f.close()\n",
    "\n",
    "r = open(options.outfile2, 'w')\n",
    "\n",
    "for record in SeqIO.parse(options.outfile, \"fasta\"):\n",
    "\tif record.id == \"Muller_A\":\n",
    "\t\tA = record\n",
    "\tif record.id == \"Muller_B\":\n",
    "\t\tB = record\n",
    "\tif record.id == \"Muller_C\":\n",
    "                C = record\n",
    "\tif record.id == \"Muller_D\":\n",
    "\t\tD = record\n",
    "\tif record.id == \"Muller_E\":\n",
    "                E = record\n",
    "\tif record.id == \"Muller_F\":\n",
    "                F = record\n",
    "\n",
    "\n",
    "\n",
    "r.write(\">\" + str(A.id) + '\\n' + str(A.seq) + '\\n')\n",
    "r.write(\">\" + str(B.id) + '\\n' + str(B.seq) + '\\n')\n",
    "r.write(\">\" + str(C.id) + '\\n' + str(C.seq) + '\\n')   \n",
    "r.write(\">\" + str(D.id) + '\\n' + str(D.seq) + '\\n')    \n",
    "r.write(\">\" + str(E.id) + '\\n' + str(E.seq) + '\\n')\n",
    "r.write(\">\" + str(F.id) + '\\n' + str(F.seq) + '\\n')\n",
    "r.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MUMmer\n",
    "Create dotplot comparison between _D.triauraria_ and _D. melanogaster_ genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "\n",
    "#SBATCH --partition=main    # Partition (job queue)\n",
    "#SBATCH --job-name=mummer         # Assign an 8-character name to your job, no spaces\n",
    "#SBATCH --nodes=1                # Number of compute nodes\n",
    "#SBATCH --ntasks=1               # Processes (usually = cores) on each node\n",
    "#SBATCH --cpus-per-task=28       # Threads per process (or per core)\n",
    "#SBATCH --export=ALL             # Export you current environment settings to the job environment\n",
    "#SBATCH --time=12:00:00\n",
    "#SBATCH --mem=20G\n",
    "#SBATCH --output=/scratch/nt365/mummer/mummer_projects/dmel_dtria/slurm-%A_%a.out\n",
    "\n",
    "source activate MUMMER\n",
    "promer --mum -b 10 -l 50 --coords -p promer_mum_dtria_rcMAMB dmel-all-chromosome-r6.22_extract_ABCDEF.fasta dtria_muller_o.fasta\n",
    "\n",
    "mummerplot -p promer_mum_dtria_rcMAMB -t postscript promer_mum_dtria_rcMAMB.delta\n",
    "\n",
    "export GNUPLOT_PS_DIR=/home/nt365/pkg/gnuplot-4.6.0/term/PostScript \n",
    "gnuplot promer_mum_dtria_rcMAMB.gp\n",
    "ps2pdf promer_mum_dtria_rcMAMB.ps"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
